tokio/tests/task_id.rs

```diff
@@ -0,0 +1,287 @@
+        fn drop(&mut self) {
+            let _ = self.tx.take().unwrap().send(task::id());
+        }
+    }
+
+    let (tx, rx) = oneshot::channel();
+    let handle = tokio::spawn(MyFuture { tx: Some(tx) });
+    let id = handle.id();
+    handle.await.unwrap();
+    assert_eq!(rx.await.unwrap(), id);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn task_id_future_destructor_abort() {
+    struct MyFuture {
+        tx: Option<oneshot::Sender<Id>>,
+    }
+
+    impl Future for MyFuture {
+        type Output = ();
+
+        fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<()> {
+            Poll::Pending
+        }
+    }
+    impl Drop for MyFuture {
+        fn drop(&mut self) {
+            let _ = self.tx.take().unwrap().send(task::id());
+        }
+    }
+
+    let (tx, rx) = oneshot::channel();
+    let handle = tokio::spawn(MyFuture { tx: Some(tx) });
+    let id = handle.id();
+    handle.abort();
+    assert_eq!(rx.await.unwrap(), id);
+}
+
+#[tokio::test(flavor = "current_thread")]
+async fn task_id_output_destructor_handle_dropped_before_completion() {
+    struct MyOutput {
+        tx: Option<oneshot::Sender<Id>>,
+    }
+
+    impl Drop for MyOutput {
+        fn drop(&mut self) {
+            let _ = self.tx.take().unwrap().send(task::id());
+        }
+    }
+
+    struct MyFuture {
+        tx: Option<oneshot::Sender<Id>>,
+    }
+
+    impl Future for MyFuture {
+        type Output = MyOutput;
+
+        fn poll(mut self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+            Poll::Ready(MyOutput { tx: self.tx.take() })
+        }
+    }
+
+    let (tx, rx) = oneshot::channel();
+    let handle = tokio::spawn(MyFuture { tx: Some(tx) });
+    let id = handle.id();
+    drop(handle);
+    assert_eq!(rx.await.unwrap(), id);
+}
+
+#[tokio::test(flavor = "current_thread")]
+async fn task_id_output_destructor_handle_dropped_after_completion() {
+    struct MyOutput {
+        tx: Option<oneshot::Sender<Id>>,
+    }
+
+    impl Drop for MyOutput {
+        fn drop(&mut self) {
+            let _ = self.tx.take().unwrap().send(task::id());
+        }
+    }
+
+    struct MyFuture {
+        tx_output: Option<oneshot::Sender<Id>>,
+        tx_future: Option<oneshot::Sender<()>>,
+    }
+
+    impl Future for MyFuture {
+        type Output = MyOutput;
+
+        fn poll(mut self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+            let _ = self.tx_future.take().unwrap().send(());
+            Poll::Ready(MyOutput {
+                tx: self.tx_output.take(),
+            })
+        }
+    }
+
+    let (tx_output, rx_output) = oneshot::channel();
+    let (tx_future, rx_future) = oneshot::channel();
+    let handle = tokio::spawn(MyFuture {
+        tx_output: Some(tx_output),
+        tx_future: Some(tx_future),
+    });
+    let id = handle.id();
+    rx_future.await.unwrap();
+    drop(handle);
+    assert_eq!(rx_output.await.unwrap(), id);
```

Comment:
We can verify in the test that the output has not yet been dropped when we drop the handle.
```suggestion
    assert!(rx_output.try_recv().is_err());
    drop(handle);
    assert_eq!(rx_output.await.unwrap(), id);
```